File: P3_Mojo/data/logistic_regression/Python/logistic_regression.py
import numpy as np
import csv

class LogisticRegression:
    def __init__(self, learning_rate=0.01, num_iterations=1000):
        self.learning_rate = learning_rate
        self.num_iterations = num_iterations
        self.weights = None
        self.bias = None
        
    def leer_archivo_csv(self):
        datos = []
        etiquetas = []
        with open('assets/logistic_regression.csv', 'r') as archivo_csv:
            lector_csv = csv.reader(archivo_csv)
            #encabezados = next(lector_csv)  # Leer la primera fila como encabezados
            for fila in lector_csv:
                tmp = list(map(int, fila))  # Convertir cada elemento de la fila a entero y agregar a la lista
                datos.append(tmp[0:2]) 
                etiquetas.append(tmp[2])
        return {'datos':datos, 'etiquetas': etiquetas}
    
    def sigmoid(self, z):
        # Aplicar clipping a los valores de entrada
        z = np.clip(z, -500, 500)
        return 1 / (1 + np.exp(-z))
    
    def initialize_parameters(self, n_features):
        self.weights = np.zeros(n_features)
        self.bias = 0
    
    def fit(self, X, y):
        self.initialize_parameters(X.shape[1])
        print(X)
        for _ in range(self.num_iterations):
            linear_model = np.dot(X, self.weights) + self.bias
            y_pred = self.sigmoid(linear_model)
            
            dw = (1 / len(X)) * np.dot(X.T, (y_pred - y))
            db = (1 / len(X)) * np.sum(y_pred - y)
            
            self.weights -= self.learning_rate * dw
            self.bias -= self.learning_rate * db
    
    def predict(self, X):
        linear_model = np.dot(X, self.weights) + self.bias
        y_pred = self.sigmoid(linear_model)
        y_pred_class = [1 if pred > 0.5 else 0 for pred in y_pred]
        return y_pred_class




# Crear e entrenar el modelo
model = LogisticRegression(learning_rate=0.1, num_iterations=10000)

# Datos de entrenamiento
data = model.leer_archivo_csv()
X_train = np.array(data['datos'])
y_train = np.array(data['etiquetas'])

model.fit(X_train, y_train)

# Datos de prueba
X_test = np.array([[4,2], [500,6]])
y_pred = model.predict(X_test)

print("Predicciones:", y_pred)

File: P3_Mojo/data/logistic_regression/CPP/logistic_regression.cpp
#include <iostream>
#include <vector>
#include <cmath>

class LogisticRegression {
private:
    double learning_rate;
    int num_iterations;
    std::vector<double> weights;
    double bias;

    double sigmoid(double z) {
        return 1 / (1 + exp(-z));
    }

    void initialize_parameters(int n_features) {
        weights = std::vector<double>(n_features, 0.0);
        bias = 0.0;
    }

public:
    LogisticRegression(double learning_rate = 0.01, int num_iterations = 1000) {
        this->learning_rate = learning_rate;
        this->num_iterations = num_iterations;
        this->weights = std::vector<double>();
        this->bias = 0.0;
    }

    void fit(std::vector<std::vector<double>>& X, std::vector<int>& y) {
        initialize_parameters(X[0].size());

        for (int i = 0; i < num_iterations; i++) {
            std::vector<double> linear_model(X.size(), 0.0);
            std::vector<double> y_pred(X.size(), 0.0);

            for (int j = 0; j < X.size(); j++) {
                for (int k = 0; k < X[0].size(); k++) {
                    linear_model[j] += X[j][k] * weights[k];
                }
                linear_model[j] += bias;
                y_pred[j] = sigmoid(linear_model[j]);
            }

            std::vector<double> dw(X[0].size(), 0.0);
            double db = 0.0;

            for (int j = 0; j < X[0].size(); j++) {
                for (int k = 0; k < X.size(); k++) {
                    dw[j] += (1.0 / X.size()) * X[k][j] * (y_pred[k] - y[k]);
                }
            }

            for (int j = 0; j < X.size(); j++) {
                db += (1.0 / X.size()) * (y_pred[j] - y[j]);
            }

            for (int j = 0; j < X[0].size(); j++) {
                weights[j] -= learning_rate * dw[j];
            }
            bias -= learning_rate * db;
        }
    }

    std::vector<int> predict(std::vector<std::vector<double>>& X) {
        std::vector<int> y_pred(X.size(), 0);

        for (int i = 0; i < X.size(); i++) {
            double linear_model = 0.0;

            for (int j = 0; j < X[0].size(); j++) {
                linear_model += X[i][j] * weights[j];
            }

            linear_model += bias;
            double y_pred_prob = sigmoid(linear_model);
            y_pred[i] = (y_pred_prob > 0.5) ? 1 : 0;
        }

        return y_pred;
    }
};

int main() {
    std::vector<std::vector<double>> X_train = { {2.0, 1.0}, {1.0, 3.0}, {0.0, 2.0}, {1.0, 1.0}, {2.0, 2.0}, {4.0, 3.0}, {6.0, 5.0}, {3.0, 5.0}, {3.0, 3.0}, {6.0, 1.0} };
    std::vector<int> y_train = { 0, 0, 0, 0, 0, 1, 1, 1, 1, 1 };

    LogisticRegression model(0.1, 1000);
    model.fit(X_train, y_train);

    std::vector<std::vector<double>> X_test = { {4.0, 2.0}, {2.0, 1.0} };
    std::vector<int> y_pred = model.predict(X_test);

    std::cout << "Predicciones:";
    for (int i = 0; i < y_pred.size(); i++) {
        std::cout << " " << y_pred[i];
    }
    std::cout << std::endl;

    return 0;
}
